File: block-hash.c
#include <stdio.h>
#include <stdint.h>

#include <openssl/sha.h>

int main(int argc, char **argv) {
    uint8_t block[80];

    if (fread(block, sizeof(block), 1, stdin) != 1) {
        fputs("Could not read block header\n", stderr);
        return 1;
    }

    SHA256_CTX ctx;
    uint8_t hash[SHA256_DIGEST_LENGTH];

    SHA256_Init(&ctx);
    SHA256_Update(&ctx, block, sizeof(block));
    SHA256_Final(hash, &ctx);

    SHA256_Init(&ctx);
    SHA256_Update(&ctx, hash, sizeof(hash));
    SHA256_Final(hash, &ctx);

    for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        printf("%x", hash[SHA256_DIGEST_LENGTH - i - 1]);
    }
    putchar('\n');

    return 0;
}

File: block-hash.py
import binascii, hashlib, locale, sys

try:
    # Python 3
    block = sys.stdin.buffer.read(80)
except AttributeError:
    # Python 2
    block = sys.stdin.read(80)

if len(block) != 80:
    raise IOError("Could not read block header")

ctx = hashlib.sha256()
ctx.update(block)
hash = ctx.digest()

ctx = hashlib.sha256()
ctx.update(hash)
hash = ctx.digest()

# binascii.hexlify returns bytes, decode for Python 3
print(binascii.hexlify(hash[::-1]).decode(locale.getpreferredencoding()))
